controller/controller.go

```diff
@@ -117,6 +117,8 @@ type Controller struct {
 	nextRunAt time.Time
 	// The nextRunAtMux is for atomic updating of nextRunAt
 	nextRunAtMux sync.Mutex
+	// The provider that is used
+	Provider provider.Provider
```

Comment:
This was my initial approach in #1849 but then I've switched from adding Provider to adding NormalizeDesiredEndpoint which is AdjustEndpoints equivalent. I think for consistency all methods should be called on registry. In #1849 they are simple proxies: https://github.com/kubernetes-sigs/external-dns/pull/1849/files#diff-3cdcf094f15e7824ba9d019ebb5acc1134c560c724b8ed35b6ce5fef2202b607

Reply:
> Controller is supposed to connect Source with Provider, hence it has references to it. However, we introduced Registry at some point which acts as a "Proxy" for a Provider that "knows how to handle ownership". If you look at Registry's interface it's identical to Provider.

I know this and absolutely thought about that. I was sitting on the idea of using registry as proxy and I'd be up to it if we can make it in a similar way that is straightforward as it is right now. I'm not particularly fond of the idea of the registry TBH, as it seems we are making things harder to understand just to have a "cache". But I'll play with it and post an update.

Reply:
@linki I'm also not sure why we have two interfaces that are identical... if both registry and providers implement the same functionality, shouldn't we define the interface once?

Reply:
I think such refactor would be fine, but for separate PR

Reply:
Back then we didn't come to agreement to go this far. They had the same interface but Registry wasn't quite conceived as a "full provider" and it might be confusing. I'm fine with with combining them into one.